File: julia_set.c
// Julia set implented in c

#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <math.h>
#include <string.h>
#include <complex.h>

extern void julia_set(
    uint32_t width, // Width of the image
    uint32_t height, // Height of the image
    uint16_t max_iter, // Maximum number of iterations
    float x_min, // Minimum x value
    float x_max, // Maximum x value
    float y_min, // Minimum y value
    float y_max, // Maximum y value
    float c_real, // Real part of the complex constant
    float c_imag, // Imaginary part of the complex constant
    uint16_t *image // Output image
)
{
    // Create a complex number from the real and imaginary parts
    const double complex c = c_real + c_imag * I;

    // Calculate the step size in the x and y directions
    const double dx = (x_max - x_min) / (double)width;
    const double dy = (y_max - y_min) / (double)height;
    
    const double escape_radius_sq = 9.0; // (3.0 squared)

    for (uint32_t y = 0; y < height; y++) {
        for (uint32_t x = 0; x < width; x++) {
            // Map the pixel coordinates to the complex plane
            double real = x_min + x * dx;
            double imag = y_min + y * dy;
            double complex z = real + imag * I; // Create a complex number from real and imag

            uint16_t iter;
            for (iter = 0; iter < max_iter; iter++) {
                z = z * z + c; // Julia set iteration formula

                // Check if the point escapes to infinity
                if (cabs(z) > escape_radius_sq) {
                    break;
                }
            }

            image[y * width + x] = iter;
        }
    }
}

File: julia_set.py
# Julia set implimentation

import numpy as np
import cv2
import tqdm
import random

def julia_set(c, n, x_min, x_max, y_min, y_max, width, height):
    """Create a julia set image

    Args:
        c (complex): Complex number
        n (int): Number of iterations
        x_min (float): Minimum x value
        x_max (float): Maximum x value
        y_min (float): Minimum y value
        y_max (float): Maximum y value
        width (int): Width of image
        height (int): Height of image

    Returns:
        numpy.ndarray: Julia set image
    """
    x = np.linspace(x_min, x_max, width)
    y = np.linspace(y_min, y_max, height)
    z = x + y[:, None] * 1j
    image = np.zeros((height, width))
    for k in tqdm.tqdm(range(n)):
        z = z**2 + c
        mask = (np.abs(z) > 2) & (image == 0)
        image[mask] = k
        z[mask] = np.nan
    return -image.T

def main():
    """Main function
    """
    c = complex(-0.1, 0.65)
    n = 200
    x_min = -0.05
    x_max = 0.5
    y_min = -0.5
    y_max = 0.05
    width = 1000
    height = 1000
    while True:
        np_set = julia_set(c, n, x_min, x_max, y_min, y_max, width, height)
        # set returned is from -n to 0, so we need to convert to 0 to 255
        image = np.uint8(-np_set / n * 255)
        cv2.imshow('image', image)
        key = cv2.waitKey(1)
        if key == ord('q'):
            break
        print('c = {}'.format(c))
        c = complex(random.uniform(-0.1, 0.1), 0.65)
if __name__ == '__main__':
    main()
